README.md
# colorcolor - Javascript Function

[![npm version](https://badge.fury.io/js/colorcolor.svg)](https://badge.fury.io/js/colorcolor)
[![Build Status](https://github.com/metaloha/color2color/actions/workflows/node.js.yml/badge.svg)](https://github.com/metaloha/color2color/actions)

The `colorcolor()` function converts HEX/HEXA/HSB/HSL/HSLA/HSV/HWB/RGB/RGBA color strings to
HEX/HEXA/HSB/HSL/HSLA/HSV/HWB/RGB/RGBA color strings.

## Demo

[http://metaloha.github.io/color2color/](http://metaloha.github.io/color2color/)

## Install

`colorcolor` is available via NPM:

	npm install colorcolor

## Usage

	colorcolor( originalColor: string, [newColorType: string( hex|hexa|hsb|hsl|hsla|hsv|hwb|rgb|rgba ) = 'rgba'] )

The first argument is the original color string in HEX, HEXA, HSB, HSL, HSLA, HSV, HWB, RGB, or RGBA format, or a CSS
named color.

The second argument (optional) is which format you'd like the new color string to be in. This will always default
to `'rgba'`.

## Examples

```ts
colorcolor('#dfe') === 'rgba(221, 255, 238, 1)'
colorcolor('#036', 'rgb') === 'rgb(0, 51, 102)'
colorcolor('rgba(64,64,64,0.5)') === 'rgba(64, 64, 64, 0.5)'
colorcolor('rgba(64 64 64 / 0.5)') === 'rgba(64, 64, 64, 0.5)'
colorcolor('rgb(64,64,64)', 'hex') === '#404040'
colorcolor('#dfe', 'rgba') === 'rgba(221, 255, 238, 1)'
colorcolor('hsla(109,100%,37%,1)') === 'rgba(35, 189, 0, 1)'
colorcolor('hsla(0.35turn 70% 55% / 1)') === 'rgba(60, 221, 76, 1)'
colorcolor('rgba(35,189,0,0.75)', 'hsl') === 'hsl(109, 100%, 37%)'
colorcolor('rgba(85%,55%,10.5%,70%)', 'hsl') === 'hsl(36, 78%, 48%)'
colorcolor('#3fa796a0', 'hsla') === 'hsla(170, 45%, 45%, 0.63)'
colorcolor('hwb(200grad 50% 25% / 0.75)') === 'rgba(128, 191, 191, 0.75)'
colorcolor('AliceBlue', 'hwb') === 'hwb(208 94% 0% / 1)'
```

## Valid color formats

`colorcolor` understands a mix of CSS and non-CSS color models, and is being actively expanded. The following formats
are currently understood by `colorcolor`:

### RGB (_[read more](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#rgb_color_model)_)

Both hexadecimal and numeric notations are supported.

* **#RGB[A]**
* **#RRGGBB[AA]**
	* **R**, **G**, **B**, and **A** are hexadecimal numbers from `00` to `ff`
	* _Examples_:
		* `#ad6`
		* `#AD6e`
		* `#f43E12`
		* `#F43e12d5`
* **rgb(R,G,B)**
* **rgb(R G B)**
* **rgba(R, G, B, A)**
* **rgba(R G B / A)**
	* **R**, **G**, and **B** can be decimals from 0 to 255 or percentages from 0% to 100%
	* **A** can be a percentage from 0% to 100% or a float from 0 to 1
	* _Examples_
		* `rgb(121, 50, 89)`
		* `rgb(121 50 89)`
		* `rgba(80, 205, 40, 0.5)`
		* `rgba(80 205 40 / 0.5)`
		* `rgb(50%, 75%, 50%)`
		* `rgba(50% 75% 50% / 75%)`

### HSL (_[read more](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hsl_color_model)_)

* **hsl(H, S, L)**
* **hsl(H S L)**
* **hsla(H, S, L, A)**
* **hsla(H S L / A)**
	* **H** is an angle expressed as `deg`, `grad`, `rad`, or `turn` (degree is assumed if the number doesn't include a
	  unit)
	* **S** and **L** are percentages from 0% to 100%
	* **A** can be a percentage from 0% to 100% or a float from 0 to 1
	* _Examples_
		* `hsl(270, 100%, 50%)`
		* `hsl(0.75turn 100% 50%)`
		* `hsla(2.65grad, 100%, 50%, 0.75)`
		* `hsla(3.14rad 100% 50% / 75%)`

### HSV/HSB (_[read more](https://color.lukas-stratmann.com/color-systems/hsv.html)_)

`hsv` and `hsb` are interchangeable and mean the same thing. These are not CSS colors, but are still supported as input
and output formats.

* **hsv(H, S, V)**
* **hsb(H, S, B)**
	* **H** is an angle expressed as `deg`, `grad`, `rad`, or `turn` (degree is assumed if the number doesn't include a
	  unit)
	* **S**, **V**, and **B** are percentages from 0% to 100%
	* _Examples_
		* `hsv(0.75turn, 100%, 50%)`
		* `hsb(270, 100%, 50%)`

### HWB (_[read more](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hwb_color_model)_)

Note that this function does not accept commas, and instead of an additional `hwba` function the opacity is optional.

* **hwb(H W B)**
* **hwb(H W B / A)**
	* **H** is an angle expressed as `deg`, `grad`, `rad`, or `turn` (degree is assumed if the number doesn't include a
	  unit)
	* **W** and **B** are percentages from 0% to 100%
	* **A** can be a percentage from 0% to 100% or a float from 0 to 1
	* _Examples_
		* `hwb(270 100% 50%)`
		* `hwb(0.75turn 100% 50%)`
		* `hwb(2.65grad 100% 50% / 0.75)`
		* `hwb(3.14rad 100% 50% / 75%)`

## Testing

You can run `npm run test` or `npm run test:watch` to run the tests alone. The coverage report can be updated
with `npm run test:coverage`.

## Support

Reach out to the maintainer at one of the following places:

- [GitHub discussions](https://github.com/metaloha/color2color/discussions)
- The email which is located [in my GitHub profile](https://github.com/metaloha)

## Contributing [![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/metaloha/color2color/issues)

First off, thanks for taking the time to contribute! Contributions are what make the open-source community such an
amazing place to learn, inspire, and create. Any contributions you make will benefit everybody else and are **greatly
appreciated**.

We have set up a separate document containing our [contribution guidelines](docs/CONTRIBUTING.md).

Thank you for being involved!

## Authors & contributors

The original setup of this repository is by [Russel Porosky](https://github.com/metaloha).

For a full list of all authors and contributors,
check [the contributor's page](https://github.com/metaloha/color2color/contributors).

## License

This project is licensed under the **MIT license**.

See [LICENSE](LICENSE) for more information.


CHANGELOG.md
## v3.0.0
* added HWB support
* update converters to directly convert between formats instead of to RGBA first
* now understands all valid formats for CSS color functions
  * _Example:_ rgba(75% 50% 66% / 50%)
* code is broken into multiple files for better readability and maintainability
* greatly expanded test coverage
* greatly expanded documentation for developers
* distribution folder includes both module and ES versions

## v2.0.3
* rebuilt with TypeScript
* fixed HSB/HSL/HSLA/HSV to allow whitespace
* minimized and unminimized versions available in the `dist` folder

## v1.1.1
* fixed some rounding issues with the HSL/HSV conversions
* added HSB (identical to HSB)
* improved testing (and made it a little worse at the same time due to the lack of 1-to-1 mapping from HSL/HSV/HSB to RGB)

## v1.1.0
* refactored to more modern JS standards
* added Grunt to automate minifying and hinting

## v1.0.1
* updated all occurrences of `color2color` to `colorcolor`

## v1.0.0
* NPM and Bower installable
** package and script name is now `colorcolor` to avoid conflict

## v0.2.1
#### All the below fixes thanks to [*Kevin Jett*](https://github.com/kevjett/)
* 3 character color names now not accidentally converted to hex
* corrected HSL to RGB saturation miscalculation
* fixed incorrect hex expansion when one or more of the original hex values were single-digit

## v0.2
* added HSL, HSLA and HSV sources and targets


src/colors.ts
/**
 * This is a collection of interfaces and types for the various color models, definitions, and
 * converter methods.
 *
 * @module
 */

/**
 * Each supported color model requires a definition object for detection and normalization.
 *
 * @see {@link detectors.ColorDetectors}
 */
export interface ColorDefinition {
	/**
	 * A list of methods that convert the current color model into another one. Every entry in the {@link ColorName} enum
	 * must be present for each definition.
	 */
	converters: {
		[key in ColorName]: Converter;
	};

	/**
	 * Convert the bits from a converter method into the format used in the color string.
	 *
	 * @remarks
	 * Fractions are converted to whole numbers (ex. 0.75 becomes 191 for RGB or 75 for percentage)
	 * or hexadecimal (ex. 0.75 becomes 'bf') and rounded off. Angles will already be in degrees
	 * before this method is called.
	 *
	 * @example
	 * // convert RGBA to HEX
	 * finalize({ r: 0.5, g: 1, b: 0.5, a: 1 }) // return { r: '80', g: 'ff', b: '80', a: 1 }
	 *
	 * @param color This is the color model (from {@link ColorDefinition.normalize}) that should be converted to the final format.
	 */
	finalize(color: ColorModel): ColorModel;

	/**
	 * Convert the bits from the regex to percentages for percentage and RGB values, and degrees for angles.
	 *
	 * @remarks
	 * Bits from the regex will begin at index 1 because index 0 is the original string. Each bit
	 * can be in any format supported by the color model, such as an angle ({@link utilities.toDegrees}), a float
	 * percentage, an actual percentage string, or a number. The normalize method is responsible for
	 * converting all of those to degrees in the case of an angle, or a float percentage for everything else. If an
	 * opacity isn't included (index {@link utilities.AlphaIndex} of `bits` is empty), then it is defaulted to `1`.
	 *
	 * @example
	 * // input to {@link ColorDefinition.re} is 'rgb(128, 255, 191)'
	 * normalize(['rgb(128, 255, 191)', '128', '255', '191']) // return [0.5, 1, 0.75, 1]
	 *
	 * @param bits Index 0 is the original string, so it should be ignored by the normalize method.
	 */
	normalize(bits: string[]): ConverterTuple;

	/**
	 * This should be able to detect a CSS string for a specific color model.
	 *
	 * @remarks
	 * Many color strings support multiple formats. For example, RGBA can be written as `rgba(191, 205, 45, 0.7)`,
	 * `rgba(75%, 40%, 100%, 50%)`, `rgba(191 205 45 / 0.7)`, and `rgba(75% 40% 100% / 50%)` (and a few others).
	 * Non-CSS color strings like `hsb` and `hsv` simply use commas and have no alternative formats. The regex should
	 * also not care about extra spaces, so `hsv(75deg, 75% 65%)` and `hsv(  75deg ,75%,   65%  )` have identical
	 * results.
	 *
	 * @see [CSS named colors](https://developer.mozilla.org/en-US/docs/Web/CSS/named-color#value)
	 * @see [RGB color model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#rgb_color_model)
	 * @see [HSL color model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hsl_color_model)
	 * @see [HWB color model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hwb_color_model)
	 */
	re: RegExp;
}

/**
 * This is the shape of the color detection object. Each supported _input_ color string needs a definition that
 * contains some necessary functionality, including string detection, number normalizing, and converting.
 *
 * @see {@link detectors.ColorDetectors}
 */
export type ColorDefinitions = {
	[key in ColorName | ColorDefinitionName]: ColorDefinition;
};

/**
 * This is a combination of the target color names and internal color names.
 */
export type ColorDefinitionType = `${ColorDefinitionName | ColorName}`;

/**
 * These are the available color formats that can be converted to.
 */
export enum ColorName {
	HEX = 'hex',
	HEXA = 'hexa',
	HSB = 'hsb',
	HSL = 'hsl',
	HSLA = 'hsla',
	HSV = 'hsv',
	HWB = 'hwb',
	RGB = 'rgb',
	RGBA = 'rgba',
}

/**
 * These are the additional color formats supported as an _input_ string.
 */
export enum ColorDefinitionName {
	HEX3 = 'hex3',
	HEXA4 = 'hexa4',
}

/**
 * This is the {@link ColorName} enum converted to a type.
 */
export type ColorType = `${ColorName}`;

/**
 * This reflects the output shape of the color {@link colors.ColorDefinition.normalize} method.
 */
export type ConverterTuple = [number | string, number, number, number];

export type HexConverter = ([r, g, b]: [number, number, number]) => ColorModel;
export type HexaConverter = ([r, g, b, a]: [
	number,
	number,
	number,
	number?,
]) => ColorModel;
export type HsbConverter = ([h, s, b]: [string, number, number]) => ColorModel;
export type HslConverter = ([h, s, l]: [string, number, number]) => ColorModel;
export type HslaConverter = ([h, s, l, a]: [
	string,
	number,
	number,
	number?,
]) => ColorModel;
export type HsvConverter = ([h, s, v]: [string, number, number]) => ColorModel;
export type HwbConverter = ([h, w, b]: [string, number, number]) => ColorModel;
export type HwbaConverter = ([h, w, b, a]: [
	string,
	number,
	number,
	number?,
]) => ColorModel;
export type RgbConverter = ([r, g, b]: [number, number, number]) => ColorModel;
export type RgbaConverter = ([r, g, b, a]: [
	number,
	number,
	number,
	number?,
]) => ColorModel;

/**
 * The output of a {@link colors.ColorDefinition.converters color converter} method must be ones of these types.
 */
export type Converter =
	| HexConverter
	| HexaConverter
	| HsbConverter
	| HslConverter
	| HslaConverter
	| HsvConverter
	| HwbConverter
	| HwbaConverter
	| RgbConverter
	| RgbaConverter;

/**
 * The RGB color model defines a given color in the sRGB color space according to its red, green, and blue components.
 * An optional alpha component represents the color's transparency.
 *
 * @see [RGB Color Model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#rgb_color_model)
 */
export type HEXA = {
	/**
	 * Red is a hexadecimal number from 0x00 to 0xff.
	 */
	r: string;

	/**
	 * Red is a hexadecimal number from 0x00 to 0xff.
	 */
	g: string;

	/**
	 * Red is a hexadecimal number from 0x00 to 0xff.
	 */
	b: string;

	/**
	 * Red is a hexadecimal number from 0x00 to 0xff.
	 */
	a: string;
};

/**
 * The HSB color model defines a given color in the sRGB color space according to its hue, saturation, and brightness
 * (also known as Value in the HSV color model).
 *
 * @see [What is HSV in Design](https://www.lifewire.com/what-is-hsv-in-design-1078068)
 */
export type HSB = {
	/**
	 * Hue is a number between 0 and 360 degrees.
	 */
	h: number;

	/**
	 * Saturation is a percentage between 0 and 1.
	 */
	s: number;

	/**
	 * Brightness (or Value) is a percentage between 0 and 1.
	 */
	b: number;
};

/**
 * The HSL color model defines a given color in the sRGB color space according to its hue, saturation, and lightness. An
 * optional alpha component represents the color's transparency.
 *
 * @see [HSL Color model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hsl_color_model)
 */
export type HSLA = {
	/**
	 * Hue is a number between 0 and 360 degrees.
	 */
	h: number;

	/**
	 * Saturation is a percentage between 0 and 1.
	 */
	s: number;

	/**
	 * Lightness is a percentage between 0 and 1.
	 */
	l: number;

	/**
	 * Opacity is a percentage between 0 and 1.
	 */
	a: number;
};

/**
 * The HSV color model defines a given color in the sRGB color space according to its hue, saturation, and value (also
 * known as Brightness in the HSB color model).
 *
 * @see [What is HSV in Design](https://www.lifewire.com/what-is-hsv-in-design-1078068)
 */
export type HSV = {
	/**
	 * Hue is a number between 0 and 360 degrees.
	 */
	h: number;

	/**
	 * Saturation is a percentage between 0 and 1.
	 */
	s: number;

	/**
	 * Value (or Brightness) is a percentage between 0 and 1.
	 */
	v: number;
};

/**
 * The HWB color model defines a given color in the sRGB color space according to its hue, whiteness, and blackness. An
 * optional alpha component represents the color's transparency.
 *
 * @see [HWB Color Model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#hwb_color_model)
 */
export type HWB = {
	/**
	 * Hue is a number between 0 and 360 degrees.
	 */
	h: number;

	/**
	 * Whiteness is a percentage between 0 and 1.
	 */
	w: number;

	/**
	 * Blackness is a percentage between 0 and 1.
	 */
	b: number;

	/**
	 * Opacity is a percentage between 0 and 1.
	 */
	a: number;
};

/**
 * The RGB color model defines a given color in the sRGB color space according to its red, green, and blue components.
 * An optional alpha component represents the color's transparency.
 *
 * @see [RGB Color Model](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#rgb_color_model)
 */
export type RGBA = {
	/**
	 * Red is a percentage between 0 and 1.
	 */
	r: number;

	/**
	 * Green is a percentage between 0 and 1.
	 */
	g: number;

	/**
	 * Blue is a percentage between 0 and 1.
	 */
	b: number;

	/**
	 * Opacity is a percentage between 0 and 1.
	 */
	a: number;
};

/**
 * All color methods work with one of these models. HEX, HSL, and RGB aren't used since alpha always {@link utilities.DefaultOpacity defaults} to 1.
 */
export type ColorModel = HEXA | HSB | HSLA | HSV | HWB | RGBA;
